### ISSUE TITLE
When docker pull callback is called, tags are not all downloaded

### ISSUE THREAD

**Issue by florentvaldelievre** [apocas/dockerode#107]
1)
````
shopmedia@staging3:~$ docker images | grep kafka
registry:5000/kafka                  0.8.1.1-24              825b925e65cc        6 minutes ago       519.8 MB
registry:5000/kafka                  latest                  825b925e65cc        6 minutes ago       519.8 MB
````

2) A new version is created and pushed on registry:5000/kafka and tagged with 'latest' and '0.8.1.1-25'
We can query the private registry to make sure everything have been pushed

http://registry:5000/v1/repositories/kafka/tags
````
{
    "0.8.1.1-24": "825b925e65cce5a08de8ac7901fdeed7bc51c272b3a0bcc4bd671187b4f7f8ea",
    "0.8.1.1-25": "bfd84b3cb6d023f7d74084c4232a3471fc0d6352f39375105f58b519becfc4d3",
    "latest": "bfd84b3cb6d023f7d74084c4232a3471fc0d6352f39375105f58b519becfc4d3"
}
````
3) I launch the script

 3-1) It pulls all the tags registry:5000/kafka. I should get the latest and the new 0.8.1.1-25 tags
 3-2) It inspects the latest and its still showing version 0.8.1.1-24 even though all the tags have been pulled

C:\dev\apps\deploy>node test.js
Pulling registry:5000/kafka
Successfully Pulled registry:5000/kafka
kafka version registry:5000/kafka:latest 825b925e65cce5a08de8ac7901fdeed7bc51c272b3a0bcc4bd671187b4f7f8ea

5) If I launch the script again, the correct version is showing up
Pulling registry:5000/kafka
Successfully Pulled registry:5000/kafka
kafka version registry:5000/kafka:latest bfd84b3cb6d023f7d74084c4232a3471fc0d6352f39375105f58b519becfc4d3

Simplified code to reproduce the issue:

````

var Docker = require('dockerode');
var docker = new Docker({host: 'http://staging3', port: 4243});
var kafka = {"image": "registry:5000/kafka"};

var pull = function(target) {
	
	console.log(" Pulling "+target.image);
	docker.pull(target.image,
		function(err, stream) { 			    	   	
			if(!err) {
				console.log("Successfully Pulled "+target.image);
				handler();
			}
		}
	 );
}

var handler = function() {

	docker.getImage(kafka.image+":latest").inspect(
		function(err, data) { 
			if(!err) {
				console.log("kafka version "+  kafka.image + ":latest "+data.Id);
			}
		}
	);
}

pull(kafka);
````

Do you have any idea ?

**Comment by skozin** [comment]
@florentvaldelievre, when `isStream == true`, as in this case, the callback [is called](https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L214) with a readable stream representing output of the command. The command is not completed at this moment, it will be completed when it's output ends.

The output is a newline-delimited JSON stream with a certain format (see [Docker remote API docs](https://docs.docker.com/reference/api/docker_remote_api_v1.16/#create-an-image)).

So you need to wait for the `stream` to end. Also it is a useful practice to register a listener for stream errors too, because otherwise your app will crash when a stream error is encountered. 

This should work:

```js
var Docker = require('dockerode'),
    docker = new Docker({ socketPath: '/var/run/docker.sock' }),
    JSONStream = require('JSONStream');


function pullImage(tag, cb, onProgress)
{
  docker.pull(tag, function(err, cmdOutStream)
  {
    if (err) return cb(err);

    var parser = JSONStream.parse(),
        output = [];

    parser.on('root', onStreamEvent);
    parser.on('error', onStreamError);
    parser.on('end', cb);

    cmdOutStream.pipe(parser);

    function onStreamEvent(evt)
    {
      output.push(evt);

      if (evt.error)
        return onStreamError(evt.error, output);
      
      onProgress && onProgress(evt);
    }

    function onStreamError(err)
    {
      parser.removeListener('root', onStreamEvent);
      parser.removeListener('error', onStreamError);
      parser.removeListener('end', cb);
      cb(err);
    }

    function onStreamEnd() {
      cb(null, output);
    }
  });
}

pullImage('registry:5000/kafka:latest',
  function onCompleted(err, output)
  {
    if (err)
      return console.error('cannot pull image:', err);

    console.log('done!');

    // useful when you are not passing onProgress callback
    // console.log('full output:', output.map(JSON.stringify).join('\n'));

    checkImage();
  },
  function onProgress(evt)
  {
    console.log('pulling progress:', evt);
  }
);

function checkImage()
{
  docker.getImage("registry:5000/kafka:latest").inspect(function(err, data)
  { 
    if (err)
      return console.error('error ingpecting image: ' + err);

    console.log("kafka:latest:" + data.Id);
  );
}
```

It uses [`JSONStream` module](https://github.com/dominictarr/JSONStream) to parse progress messages. Put it in a `pull-test.js` file, run `npm install dockerode JSONStream` and then `node pull-test.js` :)

**Comment by skozin** [comment]
@apocas, maybe we should make `isStream` an option of `docker.pull()` and similar functions, to simplify their usage when the caller is not concerned with the in-progress output and only needs the end result of an operation?

**Comment by florentvaldelievre** [comment]
Really nice @skozin , thanks a lot for that ! Sorry for the response delay

**Comment by apocas** [comment]
Yeah I definitely have been hearing about this one a lot.

My idea to solve this, is similar to what it was done with stream multiplexing.
I will test this and see how the usage ends up.

Will update the status of this issue and will ping you guys about it later.

**Comment by skozin** [comment]
@florentvaldelievre, no problem :)

@apocas, thank you for considering this! I like the idea of providing a helper function similar to `container.modem.demuxStream`. I see something like this:

```js
// get just the end result

docker.pull(tag, function(err, stream) {
  docker.followProgress(err, stream, onFinished);
});

function onFinished(err, output) {
  // pull completed successfully <=> err == null
  // output is full stream body (or maybe an array of parsed JSON events)
}
```

```js
// get the end result and progress updates

docker.pull(tag, function(err, stream) {
  docker.followProgress(err, stream, onFinished, updateProgress);
});

function updateProgress(evt) {
  // evt is already parsed (plain object)
}

function onFinished(err, output) {
  // pull completed successfully <=> err == null
  // output is full stream body (or maybe an array of parsed JSON events)
}
```

**Comment by apocas** [comment]
Hi guys,

Sorry for the delay.

As been discussed, check this out:
https://github.com/apocas/dockerode/blob/master/test/docker.js#L167
and 
https://github.com/apocas/dockerode/blob/master/test/docker.js#L149

Going to publish this soon.

**Comment by apocas** [comment]
Published on v2.1.0
